x/auth/ante/ante.go

```diff
@@ -4,30 +4,43 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/x/auth/signing"
 	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	feegrantkeeper "github.com/cosmos/cosmos-sdk/x/feegrant/keeper"
 )
 
 // NewAnteHandler returns an AnteHandler that checks and increments sequence
 // numbers, checks signatures & account numbers, and deducts fees from the first
 // signer.
 func NewAnteHandler(
-	ak AccountKeeper, bankKeeper types.BankKeeper,
+	ak AccountKeeper, bankKeeper types.BankKeeper, feeGrantKeeper *feegrantkeeper.Keeper,
 	sigGasConsumer SignatureVerificationGasConsumer,
 	signModeHandler signing.SignModeHandler,
 ) sdk.AnteHandler {
-	return sdk.ChainAnteDecorators(
+	anteDecorators := []sdk.AnteDecorator{
 		NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
 		NewRejectExtensionOptionsDecorator(),
 		NewMempoolFeeDecorator(),
 		NewValidateBasicDecorator(),
 		TxTimeoutHeightDecorator{},
 		NewValidateMemoDecorator(ak),
 		NewConsumeGasForTxSizeDecorator(ak),
-		NewRejectFeeGranterDecorator(),
+	}
+
+	if feeGrantKeeper == nil {
```

Comment:
Instead let's put the conditional before array creation and just assign a feeGrantDecorator to either the reject or deduct fee grant decorator depending on this conditional. Then it can just be one array assignment but with a var for the fee grant decorator. Does that make sense?

Reply:
Actually in `NewDeductGrantedFeeDecorator` we had added the logic like normal fee deduct + feegrant fee deduct . If we set fee granter value it uses feegrant logic else it works as normal fee deduct.
we have situations like:
1.) if we have to reject feegrant we need to have `NewRejectFeeGranterDecorator` and either `NewDeductFeeDecorator` or `NewDeductGrantedFeeDecorator` (since it works as normal fee deduct as well)

2.) if we have to allow feegrant we need to have only `NewDeductGrantedFeeDecorator` 

we cannot add `NewDeductFeeDecorator` `NewDeductGrantedFeeDecorator` together for ante builder.

Reply:
Makes sense but I wonder why we don't have separate decorators for regular fees and granted fees

Reply:
May be we can move the logic from `RejectFeeGranterDecorator` to `DeductFeeDecorator` and have only two regular fees and granted fees decorator and enable any of them based on the requirement. Wdyt ?

Reply:
@amaurym what do you think? I don't really have a strong preference but prefer we move towards whatever is the more maintainable design.

Reply:
sorry for replying late.

> why we don't have separate decorators for regular fees and granted fees

Yeah, I agree with this: `if feeGrantKeeper == nil {DeductFeeDecorator} else {DeductGrantedFeeDecorator}`

Reply:
The problem I'm seeing is duplicated logic between `DeductFeeDecorator` and `DeductGrantedFeeDecorator`. I think it's a maintenance risk to duplicate so much. I would rather have a) one decorator that handles granted fees if they're enabled or b) two decorators without duplicated logic (i.e. the granted fee decorator runs before the regular `DeductFeeDecorator`). When I originally coded all this it was approach (a) and I still think that may be simplest. Sorry for the back and forth but I think we should get this right as the ante handler is one of the most critical parts of the stack.